File: examples/file_processing/log_analyzer.py
# @depyler: optimization_level = "aggressive"
# @depyler: string_strategy = "zero_copy"
from typing import Dict, List, Optional, Tuple

class LogEntry:
    """Represents a single log entry"""
    
    def __init__(self, timestamp: str, level: str, message: str) -> None:
        self.timestamp = timestamp
        self.level = level
        self.message = message
    
    def is_error(self) -> bool:
        """Check if this is an error-level log entry"""
        return self.level.upper() in ["ERROR", "ERR", "FATAL"]
    
    def is_warning(self) -> bool:
        """Check if this is a warning-level log entry"""
        return self.level.upper() in ["WARNING", "WARN"]

class LogAnalyzer:
    """Analyzes log files and extracts insights"""
    
    def __init__(self) -> None:
        self.entries: List[LogEntry] = []
    
    def parse_log_line(self, line: str) -> Optional[LogEntry]:
        """Parse a single log line into a LogEntry"""
        # Simple format: "TIMESTAMP LEVEL MESSAGE"
        # Example: "2023-01-01 10:30:45 ERROR Database connection failed"
        
        parts = line.strip().split()
        if len(parts) < 3:
            return None
        
        # Extract timestamp (assume first two parts: date and time)
        if len(parts) >= 2:
            timestamp = parts[0] + " " + parts[1]
            level = parts[2]
            message = " ".join(parts[3:])
        else:
            return None
        
        return LogEntry(timestamp, level, message)
    
    def load_from_string(self, log_content: str) -> None:
        """Load log entries from string content"""
        lines = log_content.split('\n')
        for line in lines:
            if line.strip():
                entry = self.parse_log_line(line)
                if entry:
                    self.entries.append(entry)
    
    def count_by_level(self) -> Dict[str, int]:
        """Count log entries by level"""
        counts: Dict[str, int] = {}
        
        for entry in self.entries:
            level = entry.level.upper()
            if level in counts:
                counts[level] += 1
            else:
                counts[level] = 1
        
        return counts
    
    def get_error_messages(self) -> List[str]:
        """Get all error messages"""
        errors: List[str] = []
        for entry in self.entries:
            if entry.is_error():
                errors.append(entry.message)
        return errors
    
    def find_patterns(self, pattern: str) -> List[LogEntry]:
        """Find log entries containing specific pattern"""
        matches: List[LogEntry] = []
        pattern_lower = pattern.lower()
        
        for entry in self.entries:
            if pattern_lower in entry.message.lower():
                matches.append(entry)
        
        return matches
    
    def get_hourly_stats(self) -> Dict[str, int]:
        """Get log entry counts by hour"""
        hourly_counts: Dict[str, int] = {}
        
        for entry in self.entries:
            # Extract hour from timestamp (assume format "YYYY-MM-DD HH:MM:SS")
            timestamp_parts = entry.timestamp.split()
            if len(timestamp_parts) >= 2:
                time_part = timestamp_parts[1]
                hour_parts = time_part.split(":")
                if len(hour_parts) >= 1:
                    hour = hour_parts[0]
                    if hour in hourly_counts:
                        hourly_counts[hour] += 1
                    else:
                        hourly_counts[hour] = 1
        
        return hourly_counts
    
    def get_top_error_patterns(self, limit: int = 5) -> List[Tuple[str, int]]:
        """Get most common error patterns"""
        error_messages = self.get_error_messages()
        
        # Simple pattern extraction - look for common words in error messages
        word_counts: Dict[str, int] = {}
        
        for message in error_messages:
            words = message.lower().split()
            for word in words:
                # Skip common words
                if len(word) > 3 and word not in ["the", "and", "for", "with", "from"]:
                    if word in word_counts:
                        word_counts[word] += 1
                    else:
                        word_counts[word] = 1
        
        # Sort by count and return top patterns
        sorted_patterns: List[Tuple[str, int]] = []
        for word, count in word_counts.items():
            sorted_patterns.append((word, count))
        
        # Simple sorting by count (bubble sort)
        n = len(sorted_patterns)
        for i in range(n):
            for j in range(0, n - i - 1):
                if sorted_patterns[j][1] < sorted_patterns[j + 1][1]:
                    sorted_patterns[j], sorted_patterns[j + 1] = sorted_patterns[j + 1], sorted_patterns[j]
        
        return sorted_patterns[:limit]

def analyze_log_health(log_content: str) -> Dict[str, float]:
    """Analyze overall log health metrics"""
    analyzer = LogAnalyzer()
    analyzer.load_from_string(log_content)
    
    if not analyzer.entries:
        return {"health_score": 0.0, "error_rate": 0.0, "warning_rate": 0.0}
    
    total_entries = len(analyzer.entries)
    error_count = len([e for e in analyzer.entries if e.is_error()])
    warning_count = len([e for e in analyzer.entries if e.is_warning()])
    
    error_rate = error_count / total_entries
    warning_rate = warning_count / total_entries
    
    # Simple health score: 1.0 is perfect (no errors/warnings)
    health_score = 1.0 - (error_rate * 0.8 + warning_rate * 0.3)
    if health_score < 0.0:
        health_score = 0.0
    
    return {
        "health_score": health_score,
        "error_rate": error_rate,
        "warning_rate": warning_rate
    }

File: examples/file_processing/log_analyzer.rs
#![allow(unused_imports)]
#![allow(unused_mut)]
#![allow(unused_variables)]
#![allow(unreachable_patterns)]
#![allow(unused_assignments)]
#![allow(dead_code)]
use std::collections::HashMap;
#[derive(Debug, Clone)]
pub struct ZeroDivisionError {
    message: String,
}
impl std::fmt::Display for ZeroDivisionError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "division by zero: {}", self.message)
    }
}
impl std::error::Error for ZeroDivisionError {}
impl ZeroDivisionError {
    pub fn new(message: impl Into<String>) -> Self {
        Self {
            message: message.into(),
        }
    }
}
#[doc = r" Sum type for heterogeneous dictionary values(Python fidelity)"]
#[derive(Debug, Clone, PartialEq)]
pub enum DepylerValue {
    Int(i64),
    Float(f64),
    Str(String),
    Bool(bool),
    None,
    List(Vec<DepylerValue>),
    Dict(std::collections::HashMap<String, DepylerValue>),
}
impl std::fmt::Display for DepylerValue {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            DepylerValue::Int(i) => write!(f, "{}", i),
            DepylerValue::Float(fl) => write!(f, "{}", fl),
            DepylerValue::Str(s) => write!(f, "{}", s),
            DepylerValue::Bool(b) => write!(f, "{}", b),
            DepylerValue::None => write!(f, "None"),
            DepylerValue::List(l) => write!(f, "{:?}", l),
            DepylerValue::Dict(d) => write!(f, "{:?}", d),
        }
    }
}
impl DepylerValue {
    #[doc = r" Get length of string, list, or dict"]
    pub fn len(&self) -> usize {
        match self {
            DepylerValue::Str(s) => s.len(),
            DepylerValue::List(l) => l.len(),
            DepylerValue::Dict(d) => d.len(),
            _ => 0,
        }
    }
    #[doc = r" Check if empty"]
    pub fn is_empty(&self) -> bool {
        self.len() == 0
    }
    #[doc = r" Get chars iterator for string values"]
    pub fn chars(&self) -> std::str::Chars<'_> {
        match self {
            DepylerValue::Str(s) => s.chars(),
            _ => "".chars(),
        }
    }
    #[doc = r" Insert into dict(mutates self if Dict variant)"]
    pub fn insert(&mut self, key: String, value: DepylerValue) {
        if let DepylerValue::Dict(d) = self {
            d.insert(key, value);
        }
    }
    #[doc = r" Get value from dict by key"]
    pub fn get(&self, key: &str) -> Option<&DepylerValue> {
        if let DepylerValue::Dict(d) = self {
            d.get(key)
        } else {
            Option::None
        }
    }
    #[doc = r" Check if dict contains key"]
    pub fn contains_key(&self, key: &str) -> bool {
        if let DepylerValue::Dict(d) = self {
            d.contains_key(key)
        } else {
            false
        }
    }
    #[doc = r" Convert to String"]
    pub fn to_string(&self) -> String {
        match self {
            DepylerValue::Str(s) => s.clone(),
            DepylerValue::Int(i) => i.to_string(),
            DepylerValue::Float(fl) => fl.to_string(),
            DepylerValue::Bool(b) => b.to_string(),
            DepylerValue::None => "None".to_string(),
            DepylerValue::List(l) => format!("{:?}", l),
            DepylerValue::Dict(d) => format!("{:?}", d),
        }
    }
    #[doc = r" Convert to i64"]
    pub fn to_i64(&self) -> i64 {
        match self {
            DepylerValue::Int(i) => *i,
            DepylerValue::Float(fl) => *fl as i64,
            DepylerValue::Bool(b) => {
                if *b {
                    1
                } else {
                    0
                }
            }
            DepylerValue::Str(s) => s.parse().unwrap_or(0),
            _ => 0,
        }
    }
    #[doc = r" Convert to f64"]
    pub fn to_f64(&self) -> f64 {
        match self {
            DepylerValue::Float(fl) => *fl,
            DepylerValue::Int(i) => *i as f64,
            DepylerValue::Bool(b) => {
                if *b {
                    1.0
                } else {
                    0.0
                }
            }
            DepylerValue::Str(s) => s.parse().unwrap_or(0.0),
            _ => 0.0,
        }
    }
    #[doc = r" Convert to bool"]
    pub fn to_bool(&self) -> bool {
        match self {
            DepylerValue::Bool(b) => *b,
            DepylerValue::Int(i) => *i != 0,
            DepylerValue::Float(fl) => *fl != 0.0,
            DepylerValue::Str(s) => !s.is_empty(),
            DepylerValue::List(l) => !l.is_empty(),
            DepylerValue::Dict(d) => !d.is_empty(),
            DepylerValue::None => false,
        }
    }
}
impl std::ops::Index<usize> for DepylerValue {
    type Output = DepylerValue;
    fn index(&self, idx: usize) -> &Self::Output {
        match self {
            DepylerValue::List(l) => &l[idx],
            _ => panic!("Cannot index non-list DepylerValue"),
        }
    }
}
impl std::ops::Index<&str> for DepylerValue {
    type Output = DepylerValue;
    fn index(&self, key: &str) -> &Self::Output {
        match self {
            DepylerValue::Dict(d) => d.get(key).unwrap_or(&DepylerValue::None),
            _ => panic!("Cannot index non-dict DepylerValue with string key"),
        }
    }
}
#[derive(Debug, Clone)]
pub struct LogEntry {
    pub timestamp: String,
    pub level: String,
    pub message: String,
}
impl LogEntry {
    pub fn new(timestamp: String, level: String, message: String) -> Self {
        Self {
            timestamp,
            level,
            message,
        }
    }
    pub fn is_error(&self) -> bool {
        return ["ERROR".to_string(), "ERR".to_string(), "FATAL".to_string()]
            .contains(&self.level.clone().to_uppercase());
    }
    pub fn is_warning(&self) -> bool {
        return ["WARNING".to_string(), "WARN".to_string()]
            .contains(&self.level.clone().to_uppercase());
    }
}
#[derive(Debug, Clone)]
pub struct LogAnalyzer {
    pub entries: Vec<LogEntry>,
}
impl LogAnalyzer {
    pub fn new() -> Self {
        Self {
            entries: Vec::new(),
        }
    }
    pub fn parse_log_line(&self, line: String) -> Option<LogEntry> {
        let parts = line
            .trim()
            .to_string()
            .split_whitespace()
            .map(|s| s.to_string())
            .collect::<Vec<String>>();
        if (parts.len() as i32) < 3 {
            return None;
        };
        if (parts.len() as i32) >= 2 {
            let timestamp = parts[0 as usize] + " ".to_string() + parts[1 as usize];
            let level = parts[2 as usize];
            let message = {
                let s = &parts;
                let len = s.chars().count() as isize;
                let start_idx = (3) as isize;
                let start = if start_idx < 0 {
                    (len + start_idx).max(0) as usize
                } else {
                    start_idx as usize
                };
                s.chars().skip(start).collect::<String>()
            }
            .join(" ".to_string());
        } else {
            return None;
        };
        return Some(LogEntry::new(timestamp, level, message));
    }
    pub fn load_from_string(&mut self, log_content: String) {
        let lines = log_content
            .split("\n")
            .map(|s| s.to_string())
            .collect::<Vec<String>>();
        for line in lines {
            if line.trim().to_string() {
                let entry = self.parse_log_line(line);
                if entry {
                    self.entries.push(entry);
                };
            };
        }
    }
    pub fn count_by_level(&self) -> std::collections::HashMap<String, i32> {
        let mut counts = {
            let mut map = std::collections::HashMap::new();
            map
        };
        for entry in self.entries.clone() {
            let level = entry.level.to_uppercase();
            if counts.contains_key(&level) {
                counts.insert(level, counts[level as usize] + 1);
            } else {
                counts.insert(level, 1);
            };
        }
        return counts;
    }
    pub fn get_error_messages(&self) -> Vec<String> {
        let mut errors = vec![];
        for entry in self.entries.clone() {
            if entry.is_error() {
                errors.push(entry.message);
            };
        }
        return errors;
    }
    pub fn find_patterns(&self, pattern: String) -> Vec<LogEntry> {
        let mut matches = vec![];
        let pattern_lower = pattern.to_lowercase();
        for entry in self.entries.clone() {
            if entry.message.to_lowercase().contains(&*pattern_lower) {
                matches.push(entry);
            };
        }
        return matches;
    }
    pub fn get_hourly_stats(&self) -> std::collections::HashMap<String, i32> {
        let mut hourly_counts = {
            let mut map = std::collections::HashMap::new();
            map
        };
        for entry in self.entries.clone() {
            let timestamp_parts = entry
                .timestamp
                .split_whitespace()
                .map(|s| s.to_string())
                .collect::<Vec<String>>();
            if (timestamp_parts.len() as i32) >= 2 {
                let time_part = timestamp_parts[1 as usize];
                let hour_parts = time_part
                    .split(":")
                    .map(|s| s.to_string())
                    .collect::<Vec<String>>();
                if (hour_parts.len() as i32) >= 1 {
                    let hour = hour_parts[0 as usize];
                    if hourly_counts.contains_key(&hour) {
                        hourly_counts.insert(hour, hourly_counts[hour as usize] + 1);
                    } else {
                        hourly_counts.insert(hour, 1);
                    };
                };
            };
        }
        return hourly_counts;
    }
    pub fn get_top_error_patterns(&self, limit: i32) -> Vec<(String, i32)> {
        let error_messages = self.get_error_messages();
        let mut word_counts = {
            let mut map = std::collections::HashMap::new();
            map
        };
        for message in error_messages {
            let words = message
                .to_lowercase()
                .split_whitespace()
                .map(|s| s.to_string())
                .collect::<Vec<String>>();
            for word in words {
                if (word.len() as i32) > 3
                    && ![
                        "the".to_string(),
                        "and".to_string(),
                        "for".to_string(),
                        "with".to_string(),
                        "from".to_string(),
                    ]
                    .contains(&word)
                {
                    if word_counts.contains_key(&word) {
                        word_counts.insert(word, word_counts[word as usize] + 1);
                    } else {
                        word_counts.insert(word, 1);
                    };
                };
            }
        }
        let mut sorted_patterns = vec![];
        for (word, count) in word_counts.items() {
            sorted_patterns.push((word, count));
        }
        let n = sorted_patterns.len() as i32;
        for i in 0..n {
            for j in 0..n - i - 1 {
                if sorted_patterns[j as usize][1 as usize]
                    < sorted_patterns[j + 1 as usize][1 as usize]
                {
                    {
                        let _swap_temp =
                            (sorted_patterns[j + 1 as usize], sorted_patterns[j as usize]);
                        sorted_patterns[(j) as usize] = _swap_temp.0;
                        sorted_patterns[(j + 1) as usize] = _swap_temp.1;
                    }
                };
            }
        }
        return {
            let s = &sorted_patterns;
            let len = s.chars().count() as isize;
            let stop_idx = (limit) as isize;
            let stop = if stop_idx < 0 {
                (len + stop_idx).max(0) as usize
            } else {
                stop_idx as usize
            };
            s.chars().take(stop).collect::<String>()
        };
    }
}
#[doc = "Analyze overall log health metrics"]
#[doc = " Depyler: proven to terminate"]
pub fn analyze_log_health(
    log_content: &str,
) -> Result<HashMap<String, f64>, Box<dyn std::error::Error>> {
    let mut health_score: f64 = Default::default();
    let mut analyzer = LogAnalyzer::new();
    analyzer.load_from_string(log_content);
    if !analyzer.entries {
        return Ok({
            let mut map = HashMap::new();
            map.insert("health_score".to_string(), 0.0);
            map.insert("error_rate".to_string(), 0.0);
            map.insert("warning_rate".to_string(), 0.0);
            map
        });
    }
    let _cse_temp_0 = analyzer.entries.len() as i32;
    let total_entries = _cse_temp_0;
    let _cse_temp_1 = analyzer
        .entries
        .into_iter()
        .filter(|e| {
            let e = e.clone();
            e.is_error()
        })
        .map(|e| e)
        .collect::<Vec<_>>()
        .len() as i32;
    let error_count = _cse_temp_1;
    let warning_count = _cse_temp_1;
    let _cse_temp_2 = error_count / total_entries;
    let error_rate = _cse_temp_2;
    let _cse_temp_3 = warning_count / total_entries;
    let warning_rate = _cse_temp_3;
    let _cse_temp_4 = (error_rate as f64) * 0.8;
    let _cse_temp_5 = (warning_rate as f64) * 0.3;
    let _cse_temp_6 = 1.0 - _cse_temp_4 + _cse_temp_5;
    health_score = _cse_temp_6;
    let _cse_temp_7 = health_score < 0.0;
    if _cse_temp_7 {
        health_score = 0.0;
    }
    Ok({
        let mut map = HashMap::new();
        map.insert(
            "health_score".to_string(),
            DepylerValue::Float(health_score as f64),
        );
        map.insert(
            "error_rate".to_string(),
            DepylerValue::Int(error_rate as i64),
        );
        map.insert(
            "warning_rate".to_string(),
            DepylerValue::Int(warning_rate as i64),
        );
        map
    })
}
